fairscale/experimental/nn/multiprocess_pipe.py

```diff
@@ -85,122 +42,629 @@ def dloss(input_rref: rpc.RRef, target_rref: rpc.RRef) -> rpc.RRef:
+            )
+            self.tensors[chunk][input_idx] = input
+            self.ready_cv.notify_all()
+        return phony
+
+    def wait_for(self, chunk: int) -> None:
+        """Waits until all elements of given chunk is populated in self.tensors.
+        Then it constructs self.batches[chunk] if it is not constructed yet.
+        """
+        with self.ready_cv:
+            while self.batches[chunk] is None and any(b is None for b in self.tensors[chunk]):
+                self.ready_cv.wait()
+            if self.batches[chunk] is None:
+                tensors = cast(List[Tensor], self.tensors[chunk])
+                self.batches[chunk] = Batch(tuple(tensors), chunk)
+
+    def get_batch(self, chunk: int) -> Batch:
+        batch = self.batches[chunk]
+        assert batch is not None
+        return batch
+
+
+class PartitionHandler:
+    """This class processes a single partition of the pipeline.
+    Args:
+        module_rref: RRef to the nn.Module for this partition. It should be on the local rpc worker.
+        device: The device that holds the module.
+        num_inputs: Numer of inputs to the module
+        num_outputs: Number of outputs of the module. If the module output is not a tuple (and it is a
+            single tensor), num_outputs should be None.
+        rank: The rank of the partition
+        chunks: Number of micor-batches in a mini-batch
+        checkpoint_stop:: Checkpointing is done only for the first checkpoint_stop chunks of a mini-batch.
+    """
+
+    def __init__(
+        self,
+        module_rref: rpc.RRef,
+        device: str,
+        num_inputs: int,
+        num_outputs: Optional[int],
+        rank: int,
+        chunks: int,
+        checkpoint_stop: int,
+    ) -> None:
+        self.module = module_rref.local_value()
+        self.chunks = chunks
+        self.device = torch.device(device)
+        self.checkpoint_stop = checkpoint_stop
+        self.rank = rank
+        self.num_inputs = num_inputs
+        self.num_outputs = num_outputs
+        (self.in_queue,), (self.out_queue,) = create_workers([self.device])
+
+    def local_parameter_rrefs(self) -> List[rpc.RRef]:
+        r"""
+        Create one RRef for each parameter in the given local module, and return a
+        list of RRefs.
+        """
+        return [rpc.RRef(p) for p in self.module.parameters()]
+
+    def make_pipeline_record(self, users: List[Tuple[rpc.RRef, int, int]]) -> DistributedPipelineRecord:
+        return DistributedPipelineRecord(self.device, self.rank, self.chunks, self.num_inputs, self.num_outputs, users)
+
+    def run(self, pipeline_record: DistributedPipelineRecord) -> None:
+        """Runs pipeline parallelism. It modifies the given batches in place."""
+        m = len(pipeline_record.batches)
+
+        self.stream = current_stream(self.device)
+
+        for i in range(m):
+            with record_function("feed"):
+                pipeline_record.wait_for(i)
```

Comment:
It would be easier to understand the `wait_for`/`feed logic` if `feed()` was also called from within this function. For readability it is nice to have paired APIs balanced.

Reply:
Note, feed is called from remote worker, where as wait_for is called locally to wait until feed is called. Not sure if I understand your suggestion.
Are you suggesting I get rid of forward_results and move its logic to here?